plugins/lookup/get_moid.py

```diff
@@ -0,0 +1,579 @@
+                _filter['datacenters'] = dc_moid
+                parent_key = 'parent_' + object_type + 's'
+                _filter[parent_key] = result
+                _url = self.build_url(object_type, _filter)
+                _result = await self.fetch(_url)
+                if not _result:
+                    return None
+                try:
+                    result = _result[0][object_type]
+                except (TypeError, KeyError, IndexError) as e:
+                    raise AnsibleError(to_native(e))
+            if object_path[-1] != '/' and objects_list[-1] not in result:
+                return None
+            return result
+        elif object_type == 'cluster':
+            if object_path[-1] != '/':
+                # Save object name
+                object_name = _objects_list[-1]
+            result = await self.get_objects_moid(dc_moid, _objects_list, object_type, object_name)
+            if not result:
+                return None
+            return result
+        elif object_type in ('vm', 'datastore', 'network',  'host'):
+            # Whether the user specify the path (e.g., /my_dc/vm/) and want to use the object in that path
+            if object_path[-1] != '/':
+                # Save object name
+                object_name = _objects_list.pop(-1)
+            _result = await self.get_objects_moid(dc_moid, _objects_list, object_type, object_name)
+            if not _result:
+                return None
+            elif _result and len(_result) > 1:
+                raise AnsibleError("More than one object available: [%s]."
+                                   % ", ".join(list(item['name'] for item in _result)))
+            try:
+                result = _result[0][object_type]
+            except (TypeError, KeyError, IndexError) as e:
+                raise AnsibleError(to_native(e))
+            return result
+        else:
+            raise AnsibleError('The specified object type does not exist. object_type must be one of [%s].'
+                               % ", ".join(list(INVENTORY.keys())))
+
+    async def _get_cluster_moid(self, cluster_name, folder_moid, dc_moid):
+        _filter = {}
+        _filter["datacenters"] = dc_moid
+        _filter["folders"] = folder_moid
+        if cluster_name:
+            _filter["names"] = cluster_name
+        url = self.build_url('cluster', _filter)
+        return await self.fetch(url)
+
+    async def get_objects_moid(self, dc_moid, objects_list, object_type, object_name=None):
+        _objects = objects_list.copy()
+        folder_moid = None
+        cluster_moid = None
+        rp_moid = None
+        _filter = {}
+        _filter["datacenters"] = dc_moid
+
+        # GET folders MoID
+        _result = await self.get_folder_or_rp_moid(dc_moid, _objects, 'folder')
+        if _result:
+            for folder_info in _result:
+                if folder_info['name'] == objects_list[0]:
+                    del objects_list[0]
```

Comment:
Python passes function parameter as ref and here when you modify `objects_list`, you also modify the original version that is outside of the scope of this function. This is a bit dangerous since a function can trigger an undesirable side effect somewhere else.

So it would be better to create a new list with a different explicit name, say `temp_copy`. And a tuple may also be beneficial to be sure the list stay unchanged.